### PR TITLE
Created 2016-07-18-python-automation-netmiko-and-napalm.md

### PR DESCRIPTION
**Pull request by Maikor**
This pull request has been automatically generated by prose.io.

### CODE REVIEW

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
+---
+published: false
+date: '2016-07-18 23:11 +0300'
+title: 'Python automation: netmiko and napalm'
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:4]: Keep your title catchy. The title shows up on Twitter, so to draw attention, mention IOS-XR in the title whenever possible. 
Change this to :  "Using Netmiko and Napalm with IOS-XR: Quick Look"   or something similar

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+excerpt: Brief introduction to automation with Python
+---
+{% include toc icon="table" title="IOS-XR: Ansible and Vagrant" %}
+
+
+
+## Introduction
+
+At first, let's take a look at used tools:
+
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:19]: To begin with, let's take a look the tools we intend to use:

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+[Netmiko](https://github.com/ktbyers/netmiko) - multi-vendor ssh tool for device configuration 
+
+[NAPALM](https://github.com/napalm-automation/napalm)- tool, which provides standardized API for different platforms.
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:22]: `[NAPALM](https://github.com/napalm-automation/napalm)- python based automation tool, which provides a common API for different vendor platforms.`

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+## Introduction
+
+At first, let's take a look at used tools:
+
+[Netmiko](https://github.com/ktbyers/netmiko) - multi-vendor ssh tool for device configuration 
+
+[NAPALM](https://github.com/napalm-automation/napalm)- tool, which provides standardized API for different platforms. 
+
+We will use old setup, which consist of devbox (Ubuntu instance) and rtr (IOS-XRv)  [link]({{ site.url }}/tutorials/IOSXR-Ansible).
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:24]: Show the image of the topology here for illustration.

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+{: .notice--info}  
+
+We are interesting in IOS-XRv in particular. 
+
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:28]: We are interested in our interaction with IOS-XRv in particular.

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+Let's install python modules on devbox:
+
+```shell
+sudo pip install netmiko
+
+sudo pip install napalm
+```
+
+As first step we can verify, that setup is successful.
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:39]: To begin with   OR  to start off,  we should verify that the setup was successful.

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+Python 2.7.6 (default, Jun 22 2015, 17:58:13)
+[GCC 4.8.2] on linux2
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import netmiko
+>>> import napalm
+>>>
+
+```
+
+### Example via netmiko
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:53]: Using Netmiko

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+net_connect = ConnectHandler(**cisco_ios_xrv)
+
+output = net_connect.send_command('show ip int brief')
+print(output)
+
+output = net_connect.send_config_set(['hostname my_sweet_rtr', 'commit'])
+print(output)
+
+```
+
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:80]: Show the effect of the send commands, Do a show running config using Netmiko if possible or by directly using a show running-config on XR CLI

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+Now we can proceed with more serious example. We will have separate file with router configuration. Python will open this file, separate lines and make a list of commands from it. 
+
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:82]: Now we can proceed with a more serious example. We will use a separate file with the router configuration. Open this file in python, read the lines individually and make a list of commands from it.

File: _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+device.send_command('show ip int brief')
+
+```
+### Example via NAPALM
+
+File for telemetry configuration will be used. We can change destination port from 2103 to 2109 to try diff command. 
+
+
+At current moment commands related to configuration management doesn't respond correctly, but informative commands works fine: 
+```
```
Comment (akshshar) [inline _tutorials/2016-07-18-python-automation-netmiko-and-napalm.md:157]: but informative commands work fine:

**Comment by akshshar** [comment]
@Maikor . Added comments. Mostly corrections around the grammar and some requests for a "show run" and a diagram for  the topology. Otherwise, Looks good to me! 👍

**Comment by Maikor** [comment]
Closing this PR. @akshshar  thanks for great feedback!

**Pull request closed without merging**